### ISSUE TITLE
Large perf difference between Layer Norm with fused ops between FP16 and FP32

### ISSUE THREAD

**Issue by kevinstephano** [NVIDIA/Fuser#335]
The FP16 version is running ~3.8x slower than the FP32 version.

On A100:

FP16:
```
2868573459        7179420    1070  16384     1     1   256     1     1       64         0.000         0.001                                                     NVIDIA A100 80GB PCIe (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<CudaCodeGen::__half, (int)1, (int)1>, CudaCodeGen::Tensor<…
```

FP32:
```
 2080929144        1876968    1028  16384     1     1   512     1     1       64         0.000         0.002                                                     NVIDIA A100 80GB PCIe (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<float, (int)1, (int)1>, CudaCodeGen::Tensor<float, (int)3,…
```

Fp16 Repro:

```
import torch
from nvfuser import FusionDefinition, DataType

inputs = [
    torch.randn(20480, device='cuda', dtype=torch.float16),
    torch.randn(8, 2048, 20480, device='cuda', dtype=torch.float16),
    torch.randn(20480, device='cuda', dtype=torch.float16),
    torch.randn(20480, device='cuda', dtype=torch.float16),
    torch.randn(20480, device='cuda', dtype=torch.float16),
]

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Half, is_cpu=False)
    T1 = fd.define_tensor(symbolic_sizes=[-1, -1, -1], contiguity=[True, True, True], dtype=DataType.Half, is_cpu=False)
    T2 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Half, is_cpu=False)
    T3 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Half, is_cpu=False)
    T4 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Half, is_cpu=False)
    T5 = fd.ops.broadcast_in_dim(T0, output_shape=[8, 2048, 20480], broadcast_dims=[2])
    T6 = fd.ops.cast(T1, dtype=DataType.Float)
    T7 = fd.ops.cast(T5, dtype=DataType.Float)
    T8 = fd.ops.add(T6, T7)
    T9 = fd.ops.cast(T8, dtype=DataType.Half)
    T10 = fd.ops.broadcast_in_dim(T2, output_shape=[8, 2048, 20480], broadcast_dims=[2])
    T11 = fd.ops.cast(T9, dtype=DataType.Float)
    T12 = fd.ops.cast(T10, dtype=DataType.Float)
    T13 = fd.ops.add(T11, T12)
    T14 = fd.ops.cast(T13, dtype=DataType.Half)
    T15 = fd.ops.cast(T14, dtype=DataType.Float)
    T16 = fd.ops.var(T15, axes=[2], correction=0, keepdim=False)
    T17 = fd.ops.broadcast_in_dim(T16, output_shape=[8, 2048, 1], broadcast_dims=[0, 1])
    T18 = fd.ops.sum(T15, axes=[2], keepdim=False, dtype=DataType.Null)
    T19 = fd.ops.broadcast_in_dim(T18, output_shape=[8, 2048, 1], broadcast_dims=[0, 1])
    S20 = fd.define_constant(20480.0, dtype=DataType.Double)
    S21 = fd.ops.reciprocal(S20)
    T22 = fd.ops.mul(T19, S21)
    S23 = fd.define_constant(1.00000e-12, dtype=DataType.Double)
    T24 = fd.ops.add(T17, S23)
    T25 = fd.ops.rsqrt(T24)
    T26 = fd.ops.broadcast_in_dim(T22, output_shape=[8, 2048, 20480], broadcast_dims=[0, 1, 2])
    T27 = fd.ops.cast(T14, dtype=DataType.Float)
    T28 = fd.ops.sub(T27, T26)
    T29 = fd.ops.broadcast_in_dim(T25, output_shape=[8, 2048, 20480], broadcast_dims=[0, 1, 2])
    T30 = fd.ops.mul(T28, T29)
    T31 = fd.ops.broadcast_in_dim(T4, output_shape=[8, 2048, 20480], broadcast_dims=[2])
    T32 = fd.ops.cast(T31, dtype=DataType.Float)
    T33 = fd.ops.mul(T30, T32)
    T34 = fd.ops.broadcast_in_dim(T3, output_shape=[8, 2048, 20480], broadcast_dims=[2])
    T35 = fd.ops.cast(T34, dtype=DataType.Float)
    T36 = fd.ops.add(T33, T35)
    T37 = fd.ops.cast(T36, dtype=DataType.Half)
    fd.add_output(T37)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

for _ in range(5):
    out = fd.execute(inputs)
```

Fp32 Repro:

```
import torch
from nvfuser import FusionDefinition, DataType

inputs = [
    torch.randn(20480, device='cuda', dtype=torch.float32),
    torch.randn(8, 2048, 20480, device='cuda', dtype=torch.float32),
    torch.randn(20480, device='cuda', dtype=torch.float32),
    torch.randn(20480, device='cuda', dtype=torch.float32),
    torch.randn(20480, device='cuda', dtype=torch.float32),
]

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    T1 = fd.define_tensor(symbolic_sizes=[-1, -1, -1], contiguity=[True, True, True], dtype=DataType.Float, is_cpu=False)
    T2 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    T3 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    T4 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    T5 = fd.ops.broadcast_in_dim(T0, output_shape=[8, 2048, 20480], broadcast_dims=[2])
    T6 = fd.ops.add(T1, T5)
    T7 = fd.ops.broadcast_in_dim(T2, output_shape=[8, 2048, 20480], broadcast_dims=[2])
    T8 = fd.ops.add(T6, T7)
    T9 = fd.ops.var(T8, axes=[2], correction=0, keepdim=False)
    T10 = fd.ops.broadcast_in_dim(T9, output_shape=[8, 2048, 1], broadcast_dims=[0, 1])
    T11 = fd.ops.sum(T8, axes=[2], keepdim=False, dtype=DataType.Null)
    T12 = fd.ops.broadcast_in_dim(T11, output_shape=[8, 2048, 1], broadcast_dims=[0, 1])
    S13 = fd.define_constant(20480.0, dtype=DataType.Double)
    S14 = fd.ops.reciprocal(S13)
    T15 = fd.ops.mul(T12, S14)
    S16 = fd.define_constant(1.00000e-12, dtype=DataType.Double)
    T17 = fd.ops.add(T10, S16)
    T18 = fd.ops.rsqrt(T17)
    T19 = fd.ops.broadcast_in_dim(T15, output_shape=[8, 2048, 20480], broadcast_dims=[0, 1, 2])
    T20 = fd.ops.sub(T8, T19)
    T21 = fd.ops.broadcast_in_dim(T18, output_shape=[8, 2048, 20480], broadcast_dims=[0, 1, 2])
    T22 = fd.ops.mul(T20, T21)
    T23 = fd.ops.broadcast_in_dim(T4, output_shape=[8, 2048, 20480], broadcast_dims=[2])
    T24 = fd.ops.mul(T22, T23)
    T25 = fd.ops.broadcast_in_dim(T3, output_shape=[8, 2048, 20480], broadcast_dims=[2])
    T26 = fd.ops.add(T24, T25)
    fd.add_output(T26)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

for _ in range(5):
    out = fd.execute(inputs)
```

**Comment by liqiangxl** [comment]
Two issues in  FP16 version.
(1) Scheduler didn't correctly estimate register usage (`because of (2)`) and lead to significant spills
(2) The python script has redundant type casts  `T13[fp32] --> T14[fp16] --> T15[fp32]`, and this lead to two persistent buffers in the generated code:
```
fp16 T14[80]
fp32 T15[80]
```
In FP32 version, there is only one persistent buffer.

**After addressing issue (1), the time is 2.17 ms, by comparasion, FP32 needs 2.74 ms.**
**If manually remove redundant type casts in the script,  the time is 1.38 ms.**
